x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx

```diff
@@ -0,0 +1,200 @@
+  );
+}
+
+interface Props {
+  readonly transaction: Transaction;
+  readonly location: Location;
+}
+
+interface State {
+  readonly isOpen: boolean;
+  discoverQuery: StringMap;
+}
+
+export class TransactionActionMenu extends React.Component<Props, State> {
+  public state: State = {
+    isOpen: false,
+    discoverQuery: getDiscoverQuery(this.props.transaction)
+  };
+
+  public toggle = () => {
+    this.setState(state => ({ isOpen: !state.isOpen }));
+  };
+
+  public close = () => {
+    this.setState({ isOpen: false });
+  };
+
+  public getInfraActions(transaction: Transaction) {
+    const hostName = get(transaction, 'context.system.hostname');
+    const podId = get(transaction, 'kubernetes.pod.uid');
+    const containerId = get(transaction, 'container.id');
+    const pathname = '/app/infra';
+    const time = new Date(transaction['@timestamp']).getTime();
+    const infraMetricsQuery = getInfraMetricsQuery(transaction);
+
+    return [
+      {
+        icon: 'loggingApp',
+        label: 'Show pod logs',
+        target: podId,
+        hash: `/link-to/pod-logs/${podId}`,
+        query: { time }
+      },
+
+      {
+        icon: 'loggingApp',
+        label: 'Show container logs',
+        target: containerId,
+        hash: `/link-to/container-logs/${containerId}`,
+        query: { time }
+      },
+
+      {
+        icon: 'loggingApp',
+        label: 'Show host logs',
+        target: hostName,
+        hash: `/link-to/host-logs/${hostName}`,
+        query: { time }
+      },
+
+      {
+        icon: 'infraApp',
+        label: 'Show pod metrics',
+        target: podId,
+        hash: `/link-to/pod-detail/${podId}`,
+        query: infraMetricsQuery
+      },
+
+      {
+        icon: 'infraApp',
+        label: 'Show container metrics',
+        target: containerId,
+        hash: `/link-to/container-detail/${containerId}`,
+        query: infraMetricsQuery
+      },
+
+      {
+        icon: 'infraApp',
+        label: 'Show host metrics',
+        target: hostName,
+        hash: `/link-to/host-detail/${hostName}`,
+        query: infraMetricsQuery
+      }
+    ]
+      .filter(({ target }) => Boolean(target))
+      .map(({ icon, label, hash, query }, index) => {
+        const href = getKibanaHref({
+          location,
+          pathname,
+          hash,
+          query
+        });
+
+        return (
+          <EuiContextMenuItem icon={icon} href={href} key={index}>
+            <EuiFlexGroup gutterSize="s">
+              <EuiFlexItem>
+                <EuiLink>{label}</EuiLink>
+              </EuiFlexItem>
+              <EuiFlexItem grow={false}>
+                <EuiIcon type="popout" />
+              </EuiFlexItem>
+            </EuiFlexGroup>
+          </EuiContextMenuItem>
+        );
+      });
+  }
+
+  public async componentDidMount() {
+    const discoverQuery = await setApmIndexPatternQuery(
+      this.state.discoverQuery
+    );
+    this.setState({ discoverQuery });
```

Comment:
How do you handle if `this.props.transaction` changes?

Reply:
my first fix to deal with this was to add more lifecycle methods, then it become more complicated, so then i made getQueryWithApmIndexPattern synchronous, and only do the query object building, and had to rely on each component to do the async/state management stuff, which kind of defeats the purpose of the abstraction.

So in the end i created a new component `QueryWithIndexPattern` which handles all the async, state management, and query object transform. I ended up implementing it as a FaCC so that it the final query object can be used in either case (passed as a prop in DiscoverButton, or used to build the href in TransactionActionMenu)

LMK what you think.